cmd/image-builder-db-test/main_test.go

```diff
@@ -0,0 +1,157 @@
+)
+
+const (
+	ANR1 = "000001"
+	ANR2 = "000002"
+	ANR3 = "000003"
+
+	ORGID1= "100000"
+)
+
+func conf(t *testing.T) *config.ImageBuilderConfig {
+	c := config.ImageBuilderConfig{
+		ListenAddress: "unused",
+		LogLevel:      "INFO",
+		MigrationsDir: "/usr/share/image-builder/migrations",
+		PGHost:        "localhost",
+		PGPort:        "5432",
+		PGDatabase:    "imagebuilder",
+		PGUser:        "postgres",
+		PGPassword:    "foobar",
+		PGSSLMode:     "disable",
+	}
+
+	err := config.LoadConfigFromEnv(&c)
+	require.NoError(t, err)
+
+	return &c
+}
+
+func connStr(t *testing.T) string {
+	c := conf(t)
+	return fmt.Sprintf("postgres://%s:%s@%s:%s/%s?sslmode=%s", c.PGUser, c.PGPassword, c.PGHost, c.PGPort, c.PGDatabase, c.PGSSLMode)
+}
+
+func migrateOneStep(t *testing.T) {
+	c := conf(t)
+
+	log, err := logger.NewLogger(c.LogLevel, nil, nil, nil, nil)
+	require.NoError(t, err)
+	require.NotNil(t, log)
+
+	err = db.MigrateSteps(connStr(t), c.MigrationsDir, 1, log)
+	require.NoError(t, err)
+}
+
+func migrateUp(t *testing.T) {
+	c := conf(t)
+
+	log, err := logger.NewLogger(c.LogLevel, nil, nil, nil, nil)
+	require.NoError(t, err)
+	require.NotNil(t, log)
+
+	err = db.Migrate(connStr(t), c.MigrationsDir, log)
+	require.NoError(t, err)
+}
+
+func connect(t *testing.T) *pgx.Conn {
+	conn, err := pgx.Connect(context.Background(), connStr(t))
+	require.NoError(t, err)
+	return conn
+}
+
+// after each migration insert some data which follows the schema of that migration
+func insertData1(t *testing.T) {
+	conn := connect(t)
+	defer conn.Close(context.Background())
+
+	insert := "INSERT INTO composes(job_id, request, created_at, account_id, org_id) VALUES ($1, $2, CURRENT_TIMESTAMP, $3, $4)"
+
+	_, err := conn.Exec(context.Background(), insert, uuid.New().String(), "{}", ANR1, ORGID1)
+	require.NoError(t, err)
+
+	_, err = conn.Exec(context.Background(), insert, uuid.New().String(), "{}", ANR1, ORGID1)
+	require.NoError(t, err)
+
+	_, err = conn.Exec(context.Background(), insert, uuid.New().String(), "{}", ANR2, ORGID1)
+	require.NoError(t, err)
+}
+
+func insertData2(t *testing.T) {
+	conn := connect(t)
+	defer conn.Close(context.Background())
+
+	insert := "INSERT INTO composes(job_id, request, created_at, account_number, org_id) VALUES ($1, $2, CURRENT_TIMESTAMP, $3, $4)"
+
+	_, err := conn.Exec(context.Background(), insert, uuid.New().String(), "{}", ANR1, ORGID1)
+	require.NoError(t, err)
+
+	_, err = conn.Exec(context.Background(), insert, uuid.New().String(), "{}", ANR1, ORGID1)
+	require.NoError(t, err)
+
+	_, err = conn.Exec(context.Background(), insert, uuid.New().String(), "{}", ANR2, ORGID1)
+	require.NoError(t, err)
+}
+
+func TestMain(t *testing.T) {
+	migrateOneStep(t)
+	insertData1(t)
+
+	migrateOneStep(t)
+	insertData2(t)
+
+	// make sure migrating a fully migrated db doesn't error out
+	migrateUp(t)
+
+	d, err := db.InitDBConnectionPool(connStr(t))
+	require.NoError(t, err)
+
+	composes, count, err := d.GetComposes(ANR1, 100, 0)
```

Comment:
Thinking about this part, to ease future amendments, I'd define tests scenarios with their own functions.

```
test_migration()
test_count_returns_total_in_db()
test_selected_compose_equals_inserted_compose()
test_no_cross_account_access()
test_compose_insertion()
```

what are your thoughts ?

Reply:
Depends what you mean; 

If it's introducing more helpers, I think it's too little for now to split up into more helper functions.

If you mean to make more Test* functions: we can't rely on the order of unit tests,  and those should be able to be executed in parallel. So we can really only have one Test* function per database, and there's only one database atm. It's possible I suppose if we move the podman run call into the test itself, and do the migration per test.

Reply:
Yest, I see it as 'introducing more helpers'. Just to keep track of the meaning of each steps of the test.

Reply:
But yeah also, as you said, this is a simple test, so it might be too much splitting, your call :)

Reply:
Right, not a huge fan I think because the state of the DB would change in between those helpers, and the helpers would read as subtests. I think it's fine to split it up more if each unit test would get it's own database. But let's do that in a followup then.

I added some comments in stead. And then let's [make the test invoke podman](https://podman.io/blogs/2020/08/10/podman-go-bindings.html) in a future iteration.

Reply:
I completely agree.